torch/csrc/jit/tensorexpr/graph_opt.cpp

```diff
@@ -0,0 +1,170 @@
+#include <torch/csrc/jit/tensorexpr/graph_opt.h>
+
+#include <torch/csrc/jit/passes/tensorexpr_fuser.h>
+#include <torch/csrc/jit/tensorexpr/kernel.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+// Move the given user of `aten::cat` op to its inputs.
+Node* moveCatAfterUse(Node* cat, Node* user, std::shared_ptr<Graph> subgraph) {
+  // Example IR:
+  //   %1 = ...
+  //   %2 = ...
+  //   %3 = prim::ListConstruct(%1, %2)
+  //   %4 = aten::cat(%3, ...)
+  //   %5 = aten::relu(%4)
+  //   return (%5)
+  //
+  // To be transformed to:
+  //   %1 = ...
+  //   %2 = ...
+  //   %5.1 = aten::relu(%1)
+  //   %5.2 = aten::relu(%2)
+  //   %3 = prim::ListConstruct(%5.1, %5.2)
+  //   %4 = aten::cat(%3, ...)
+  //   return (%4)
+
+  TORCH_INTERNAL_ASSERT(cat->output()->hasUses());
+  TORCH_INTERNAL_ASSERT(cat->output()->uses().size() == 1);
+  TORCH_INTERNAL_ASSERT(cat->input(0)->node()->kind() == prim::ListConstruct);
+  auto cat_list = cat->input(0)->node();
+  auto cat_inputs = cat_list->inputs();
+
+  auto user_tensor_type = user->output()->type()->cast<c10::TensorType>();
+  TORCH_INTERNAL_ASSERT(user_tensor_type);
+  std::unordered_map<Value*, Value*> new_cat_inputs;
+  for (auto inp : cat_inputs) {
+    auto new_cat_input = subgraph->createClone(
+        user, [&](Value* k) { return (k == cat->output()) ? inp : k; });
+    // The type of new input should be same as the type of the user. This is
```

Comment:
By `type` you mean specifically `scalar type` here, right, since obviously the size of the input and the size of the user will be different?

Reply:
This comment was left over from some earlier changes I was making. Removed now.